sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -76,41 +98,129 @@ private static void CheckForRedundantConditions(ExplodedGraph explodedGraph, Syn
+            if (statement != null)
+            {
+                yield return statement.GetLocation();
+            }
+        }
+
+        private static TextSpan GetSpan(SyntaxNode startNode, SyntaxNode endNode)
         {
-            foreach (var alwaysTrue in conditionTrue.Except(conditionFalse))
+            if (startNode.Equals(endNode))
             {
-                context.ReportDiagnostic(Diagnostic.Create(rule, alwaysTrue.GetLocation(), "true"));
+                return startNode.Span;
             }
+            return TextSpan.FromBounds(startNode.SpanStart, endNode.Span.End);
+        }
 
-            foreach (var alwaysFalse in conditionFalse.Except(conditionTrue))
+        private static SyntaxNode GetUnreachableStatement(IEnumerable<ExpressionSyntax> unreachableExpressions, bool constantValue)
+        {
+            var parent = unreachableExpressions
+                .Select(SyntaxHelper.GetSelfOrTopParenthesizedExpression) // unreachable expressions
+                .Select(node => node.Parent is BinaryExpressionSyntax
+                    ? node.Parent.Parent
+                    : node.Parent) // Constant node is the only expression in an if statement condition
+                .FirstOrDefault();
+
+            if (constantValue)
+            {
+                return (SyntaxNode)
+                    (parent as ConditionalExpressionSyntax)?.WhenFalse ??
+                    (parent as IfStatementSyntax)?.Else?.Statement;
+            }
+            else
             {
-                context.ReportDiagnostic(Diagnostic.Create(rule, alwaysFalse.GetLocation(), "false"));
+                return (SyntaxNode)
+                    (parent as ConditionalExpressionSyntax)?.WhenTrue ??
+                    (parent as IfStatementSyntax)?.Statement ??
+                    (parent as WhileStatementSyntax)?.Statement ??
+                    (parent as DoStatementSyntax)?.Statement;
             }
         }
 
+        private static IEnumerable<ExpressionSyntax> GetUnreachableExpressions(SyntaxNode constantExpression, bool constantValue)
+        {
+            foreach (var current in constantExpression.AncestorsAndSelf())
+            {
+                if (current.Parent is ParenthesizedExpressionSyntax)
+                {
+                    continue;
+                }
+
+                var binary = current.Parent as BinaryExpressionSyntax;
+                if (!IsShortcuttingExpression(binary, constantValue))
+                {
+                    break;
+                }
+
+                if (binary.Left == current)
+                {
+                    yield return binary.Right;
+                }
+            }
+        }
+
+        private static bool IsShortcuttingExpression(BinaryExpressionSyntax expression, bool constantValueIsTrue)
+        {
+            return expression != null &&
+                (expression.Kind() == SyntaxKind.LogicalAndExpression && !constantValueIsTrue ||

```

Comment:
I tend to prefer using the `.IsKind(SyntaxKind.XXX)` instead of equality on Kind().